BFB restartability for mpas-ocean and -seaice

dazlich opened the pull request:
This modification lets the user get the BFBrestartability branches for mpas-ocean and mpas-seaice, which have established bit-for-bit restartability for the restart files and some of the diagnostic output for those components.

gdicker1 commented:
Sub-PRs:
- [ ] https://github.com/EarthWorksOrg/mpas-ocean/pull/19
- [ ] https://github.com/EarthWorksOrg/mpas-seaice/pull/21

gdicker1 commented:
@dazlich, could you extend the PR description to be a little more specific? Maybe something like "... and the diagnostic output except for daily and monthly time-averaged fields."

I'll use this for the merge commit message and want that to be a little more clear.

dazlich commented:
Specifically, time-averaged diagnostics (timeSeriesStatsDaily and timeSeriesStatsMonthly) files fail to accumulate the first timestep values when constructing their means, and thus their diagnostic output fails to be reproducible when restarting. Work is continuing to diagnose and fix this. Other diagnostic output is now reproducible by fixing default namelist values.

dazlich commented:
Hold on a bit, I am doing some more testing and see some issues in the ocean. At first I thought it was a problem with some hi-res input, but now I am seeing it with nvhpc at 120km.

dazlich commented:
To be more specific, mpas-ocean does a positive-definite check on the imported swdwn field and I am seeing failures at some resolutions and certain configurations. Still trying to nail this down. I am also seeing this failure with some 120km gpu tests a looking for possible connections to #75

dazlich commented:
Ok, I've got this sorted out. The additional ocn_import call that I added for bit-for-bit restartability didn't have all the fluxes defined at that point. This messed up the check for positive SW flux in some configurations. I've modified some ocean code so this check is bypassed for the initial import call. This is ready for review now.

gdicker1 commented:
There was a problem I found when testing this PR. Pasting again here for visibility

> I think this may need some more looking. I could only get Intel to run when I tried these changes. I used create_test for exact restart tests using the CHAOS2000dev compset at 120km using the Intel, NVHPC, and GNU compilers. Both NVHPC and GNU appear to abort during init, before the first timestep...

See [this comment in EarthWorksOrg/mpas-ocean #19](https://github.com/EarthWorksOrg/mpas-ocean/pull/19#issuecomment-2606107287) for more.

dazlich commented:
If I comment out the shortwave test nvhpc works. But gnu still has a problem in ocn_export that I am working to track down. I think the shortwave test is superfluous and may leave it commented for this PR.

dazlich commented:
I think gnu is testing an errorcode that never gets properly initialized - testing a fix.

I am rewriting the shortwave test to avoid the any statement.

dazlich commented:
@gdicker1, @areanddee - Ok, nvhpc and gnu compilers now work with this PR and it is ready for review (again).

Gnu failed on the test of an errorcode that had not been initialized. Possibly more of them in the seaice and ocean nuopc caps and I will track them down in the near future.

Nvhpc failed testing for positive shortwave. It apparently had problems with the fortran ‘ANY’ function. Rewriting this test fixed the problem. The old test remains commented if the nvidia people want to have a look at it.

gdicker1 commented:
@dazlich, I'm waiting for a final GNU run to go through; it died from out of memory before it created both restart files I wanted to compare. I'll look into that memory, but I don't think it's related to these changes.

gdicker1 commented:
@dazlich, could I get some reassurance that CHAOS2000dev results look reasonable to you? If the results seem reasonable I think this can be approved.

I'm seeing lots of differences from the previous tag's baselines but they're all pretty small. Very few of the RMS NORMALIZED values reported in the \*.cprnc.out files are on the order of E+01 and all the rest are smaller. Just many many fields affected (as expected for a coupled system). I'm also not seeing exact matches for "COMPARE_base_rest" steps for ERS CHAOS2000dev tests, but I think that's expected since not _every_ problem field is addressed by this PR.

My `create_test` runs are in sub-dirs of `"/glade/derecho/scratch/gdicker/ew_PR99_2025Jan23193000"` and are compared against results in `"/glade/campaign/univ/ucsu0085/ew_test_baselines/ewm-2.4.001/"`

dazlich commented:
@gdicker1 - I’ve had a look at several of the atm fields in the 9 timestep runs and they all look reasonable

gdicker1 reviewed (approved):
Thank you. This will be merged as ewm-2.4.002 this afternoon.
